SymCryptProvider/src/kem/p_scossl_mlkem.c

```
@@ -0,0 +1,524 @@
+        ctx->provCtx = provctx;
+    }
+
+    return ctx;
+}
+
+static void p_scossl_mlkem_freectx(_Inout_ SCOSSL_MLKEM_CTX *ctx)
+{
+    if (ctx == NULL)
+        return;
+
+    p_scossl_ecdh_freectx(ctx->classicKeyexchCtx);
+    OPENSSL_free(ctx);
+}
+
+static SCOSSL_MLKEM_CTX *p_scossl_mlkem_dupctx(_In_ SCOSSL_MLKEM_CTX *ctx)
+{
+    SCOSSL_MLKEM_CTX *copyCtx = OPENSSL_malloc(sizeof(SCOSSL_MLKEM_CTX));
+
+    if (copyCtx != NULL)
+    {
+        copyCtx->keyCtx = ctx->keyCtx;
+        copyCtx->operation = ctx->operation;
+        copyCtx->provCtx = ctx->provCtx;
+
+        if (ctx->classicKeyexchCtx != NULL)
+        {
+            copyCtx->classicKeyexchCtx = NULL;
+        }
+        else if ((copyCtx->classicKeyexchCtx = p_scossl_ecdh_dupctx(ctx->classicKeyexchCtx)) == NULL)
+        {
+            OPENSSL_free(copyCtx);
+            copyCtx = NULL;
+        }
+    }
+
+    return copyCtx;
+}
+
+static SCOSSL_STATUS p_scossl_mlkem_classic_keyexch_init(_Inout_ SCOSSL_MLKEM_CTX *ctx, _In_ SCOSSL_ECC_KEY_CTX *classicKeyCtx)
+{
+    if (ctx->classicKeyexchCtx == NULL &&
+        (ctx->classicKeyexchCtx = p_scossl_ecdh_newctx(ctx->provCtx)) == NULL)
+    {
+        ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+        return SCOSSL_FAILURE;
+    }
+
+    if (p_scossl_ecdh_init(ctx->classicKeyexchCtx, classicKeyCtx, NULL) != SCOSSL_SUCCESS)
+    {
+        return SCOSSL_FAILURE;
+    }
+
+    return SCOSSL_SUCCESS;
+}
+
+static SCOSSL_STATUS p_scossl_mlkem_init(_Inout_ SCOSSL_MLKEM_CTX *ctx, _In_ SCOSSL_MLKEM_KEY_CTX *keyCtx,
+                                         int operation)
+{
+    if (ctx == NULL || keyCtx == NULL)
+    {
+        ERR_raise(ERR_LIB_PROV, ERR_R_PASSED_NULL_PARAMETER);
+        return SCOSSL_FAILURE;
+    }
+
+    if (keyCtx->key == NULL ||
+        keyCtx->groupInfo == NULL)
+    {
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY);
+        return SCOSSL_FAILURE;
+    }
+
+    ctx->keyCtx = keyCtx;
+    ctx->operation = operation;
```

>ctx->operation = operation; [](http://example.com/codeflow?start=4&length=27)

One could fail early here if we're setting up a decapsulation context for an ML-KEM key context which only has the encapsulation key (rather than fail in the call to decapsulate).

Could be a source of minor functional differences between providers.

---

Looks like this is what OpenSSL is doing in 3.5 so I'll add this check.